### PR TITLE
Add switch for notification

### PR DESCRIPTION
**Pull request by Gentleelephant**
Signed-off-by: Gentleelephant <birdhk@kubesphere.io>

### What type of PR is this?
/kind feature

### What this PR does / why we need it:
Support for pluggable notification
### Which issue(s) this PR fixes:

Fixes #

### Special notes for reviewers:
```
```

### Does this PR introduced a user-facing change?

```release-note
None
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -507,17 +515,19 @@ func (s *APIServer) waitForResourceSync(ctx context.Context) error {
 		{Group: "network.kubesphere.io", Version: "v1alpha1"}: {
 			"ippools",
 		},
-		{Group: "notification.kubesphere.io", Version: "v2beta1"}: {
+	}
+	if s.Config.MultiClusterOptions.Enable {
```
Comment (wansir) [inline pkg/apiserver/apiserver.go:519]: Notification is controlled by `Config.MultiClusterOptions.Enable`, right?

Comment (Gentleelephant) [reply pkg/apiserver/apiserver.go:519]: I refer to the controller's code:
https://github.com/kubesphere/kubesphere/blob/master/cmd/controller-manager/app/controllers.go#L535

Comment (wansir) [reply pkg/apiserver/apiserver.go]: It's not a good choice,  the API needs a separate control flag, and the notification controller has a different logic.

Comment (Gentleelephant) [reply pkg/apiserver/apiserver.go]: I redefined a variable that controls whether notifications are turned on or not.
image

File: pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -213,7 +214,8 @@ func (s *APIServer) installMetricsAPI() {
 
 // Install all kubesphere api groups
 // Installation happens before all informers start to cache objects, so
-//   any attempt to list objects using listers will get empty results.
+//
```
Comment (wansir) [inline pkg/apiserver/apiserver.go:217]:
```suggestion
```
Remove the useless changes.

File: pkg/simple/client/notification/options.go [hunk 3]
```diff
@@ -18,11 +18,14 @@ package notification
 
 type Options struct {
 	Endpoint string `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+
+	Enable bool `json:"enable,omitempty" yaml:"enable,omitempty"`
```
Comment (wansir) [inline pkg/simple/client/notification/options.go:22]: It's similar to Auditing and Events, empty endpoint means disabled.

File: pkg/apiserver/apiserver.go [hunk 4]
```diff
@@ -213,7 +214,7 @@ func (s *APIServer) installMetricsAPI() {
 
 // Install all kubesphere api groups
 // Installation happens before all informers start to cache objects, so
-//   any attempt to list objects using listers will get empty results.
+//	any attempt to list objects using listers will get empty results.
```
Comment (wansir) [inline pkg/apiserver/apiserver.go:217]:
```suggestion
//   any attempt to list objects using listers will get empty results.
```
Same here.

Comment (Gentleelephant) [reply pkg/apiserver/apiserver.go]: Sorry, there should be a problem with some indentation of the computer.

File: pkg/simple/client/notification/options.go [hunk 5]
```diff
@@ -18,11 +18,14 @@ package notification
 
 type Options struct {
 	Endpoint string `json:"endpoint,omitempty" yaml:"endpoint,omitempty"`
+
+	Enable bool `json:"enable" yaml:"enable"`
```
Comment (wansir) [inline pkg/simple/client/notification/options.go:22]:
```suggestion
```

File: pkg/apiserver/apiserver.go [hunk 6]
```diff
@@ -261,18 +262,20 @@ func (s *APIServer) installKubeSphereAPIs(stopCh <-chan struct{}) {
...
 		s.KubernetesClient.Prometheus(), s.AlertingClient, s.Config.AlertingOptions))
 	urlruntime.Must(alertingv2beta1.AddToContainer(s.container, s.InformerFactory, s.AlertingClient))
 	urlruntime.Must(version.AddToContainer(s.container, s.KubernetesClient.Kubernetes().Discovery()))
 	urlruntime.Must(kubeedgev1alpha1.AddToContainer(s.container, s.Config.KubeEdgeOptions.Endpoint))
 	urlruntime.Must(edgeruntimev1alpha1.AddToContainer(s.container, s.Config.EdgeRuntimeOptions.Endpoint))
-	urlruntime.Must(notificationkapisv2beta1.AddToContainer(s.container, s.InformerFactory, s.KubernetesClient.Kubernetes(),
-		s.KubernetesClient.KubeSphere()))
-	urlruntime.Must(notificationkapisv2beta2.AddToContainer(s.container, s.InformerFactory, s.KubernetesClient.Kubernetes(),
-		s.KubernetesClient.KubeSphere(), s.Config.NotificationOptions))
+	if s.Config.NotificationOptions.Enable {
```
Comment (wansir) [inline pkg/apiserver/apiserver.go:272]:
```suggestion
	if s.Config.NotificationOptions.isEnabled() {
```

File: pkg/simple/client/notification/options.go (hunk 5 above)
Comment (Gentleelephant) [reply pkg/simple/client/notification/options.go:22]: delete enable ?

Comment (Gentleelephant) [reply pkg/simple/client/notification/options.go:22]: If enable is removed, how can the user modify it through the configuration file? @wansir

Comment (Gentleelephant) [reply pkg/simple/client/notification/options.go:22]: If the user has configured the endpoint, then notification is enabled?

Comment (wansir) [reply pkg/simple/client/notification/options.go:22]: yeap

Comment (wansir) [reply pkg/simple/client/notification/options.go:22]: It's similar to Auditing and Events.

**Pull request merged**